### PR TITLE
Vscode integration again

### PR DESCRIPTION
**Pull request by Pesc0**
follows #9663 
keep discussion centralized there

### CODE REVIEW

**Comment by Pesc0** [comment]
@adrianinsaval  ready for merge

**Comment by adrianinsaval** [comment]
great, the idea here is that people using vscode run `cp contrib/.vscode .vscode` right? we probably want to add a readme file explaining what all this is for and how to update the environment when we update it here, maybe including basic install instructions for mambaforge. This can be a follow up PR of course.

maybe we should move the conda folder to contrib too. Or to tools/build?

just to know, what's wrong with the windows debugger?

testing on windows it's failing to compile fore me, do we know anything about this?
```
[build] [743/5314] Linking CXX shared library bin\FreeCADBase_d.dll
[build] FAILED: bin/FreeCADBase_d.dll src/Base/FreeCADBase_d.lib 
[build] cmd.exe /C "cd . && C:\Users\adrian\Documents\repos\freecad\FreeCAD\.conda\freecad\Library\bin\cmake.exe -E vs_link_dll --intdir=src\Base\CMakeFiles\FreeCADBase.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~3\2019\BUILDT~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\FreeCADBase.rsp  /out:bin\FreeCADBase_d.dll /implib:src\Base\FreeCADBase_d.lib /pdb:bin\FreeCADBase_d.pdb /dll /version:0.0 /machine:x64 /NODEFAULTLIB /debug /INCREMENTAL  && cd ."
[build] LINK Pass 1: command "C:\PROGRA~2\MICROS~3\2019\BUILDT~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\FreeCADBase.rsp /out:bin\FreeCADBase_d.dll /implib:src\Base\FreeCADBase_d.lib /pdb:bin\FreeCADBase_d.pdb /dll /version:0.0 /machine:x64 /NODEFAULTLIB /debug /INCREMENTAL /MANIFEST /MANIFESTFILE:src\Base\CMakeFiles\FreeCADBase.dir/intermediate.manifest src\Base\CMakeFiles\FreeCADBase.dir/manifest.res" failed (exit code 1120) with the following output:
[build] LINK : warning LNK4044: unrecognized option '/lpthreads'; ignored
[build]    Creating library src\Base\FreeCADBase_d.lib and object src\Base\FreeCADBase_d.exp
[build] UnitPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] PyTools.c.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] TypePyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] VectorPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] UnitsApiPy.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] QuantityPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] RotationPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] Stream.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] swigpyrun.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] Tools.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] PersistencePyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] PlacementPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] PrecisionPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] PyObjectBase.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] Exception.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] GeometryPyCXX.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] Interpreter.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] MatrixPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] BoundBoxPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] Console.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] ConsoleObserver.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] CoordinateSystemPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] IndirectPythonInterface.cxx.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] AxisPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] BaseClass.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] BaseClassPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_NegativeRefcount
[build] UnitPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] PyTools.c.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] TypePyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] VectorPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] UnitsApiPy.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] QuantityPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] RotationPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] Stream.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] swigpyrun.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] Tools.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] PersistencePyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] PlacementPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] PrecisionPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] PyObjectBase.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] Exception.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] GeometryPyCXX.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] Interpreter.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] MatrixPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] BoundBoxPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] Console.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] ConsoleObserver.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] CoordinateSystemPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] IndirectPythonInterface.cxx.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] AxisPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] BaseClass.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] BaseClassPyImp.cpp.obj : error LNK2001: unresolved external symbol __imp__Py_RefTotal
[build] bin\FreeCADBase_d.dll : fatal error LNK1120: 2 unresolved externals
```

**Comment by Pesc0** [comment]
> great, the idea here is that people using vscode run cp contrib/.vscode .vscode right? 

correct

> we probably want to add a readme file explaining what all this is for 

I will add a section in the developers handbook after this gets merged.

> maybe we should move the conda folder to contrib too. Or to tools/build?

This question is for @oursland 

> just to know, what's wrong with the windows debugger?

It is a bit tricky to get working. also it is of "type": "cppvsdbg", so we can't just add it in the current debugging configuration. it would require a new, separate profile. I postponed finding a solution for this to get the current changes merged first.

> testing on windows it's failing to compile fore me, do we know anything about this?

Hm that's weird. Compile should work just fine. Is it a conda problem? Can you compile with libpack? (select the non-conda preset)

**Comment by oursland** [comment]
> great, the idea here is that people using vscode run `cp contrib/.vscode .vscode` right? we probably want to add a readme file explaining what all this is for and how to update the environment when we update it here, maybe including basic install instructions for mambaforge. This can be a follow up PR of course.

Correct, from my perspective as well.

> maybe we should move the conda folder to contrib too. Or to tools/build?

We can move it somewhere, but I do intend on making this a bit of a first-class part of the system to match the way official builds are made in a cross-platform way.

I am working on getting the CI system updated for both the presets and to support conda builds on all three platforms.  Progress can be found on my [fork on the `ci-update` branch](https://github.com/oursland/FreeCAD/tree/ci-update), and a PR will be made when it is worth reviewing.

When this is fully functional for builds and testing, I will move into documenting the approach for local development and testing via conda.

> just to know, what's wrong with the windows debugger?

I have had no luck in getting the debugger (any debugger) to launch FreeCAD.exe.  You can, however, get the debuggers to attach to a running FreeCAD.exe instance.  I suspect the issue is that `FreeCAD.exe` forks and exits on startup, causing the program to exit and return to the terminal.  This is different behavior than on Linux and macOS where running `FreeCAD` retains the same PID and does not return until the program terminates.

> testing on windows it's failing to compile fore me, do we know anything about this?

This is a conda-specific issue.  The package `python` does not provide the appropriate debug libraries for the python interpreter on Windows, resulting in failed linking.  You can build the `conda-windows-release` preset for that platform.

File: contrib/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        "MIMode": "lldb"
+      },
+      "stopAtEntry": false,
+      "externalConsole": false,
+      "presentation": {
+        "hidden": true,
+      }
+    },
+    {
+      "name": "Python debugger",
```
Comment (oursland) [inline contrib/.vscode/launch.json:62]: I will again raise my disagreement on this approach.   FreeCAD has built-in support for launching the VS Code debugger protocol, so having a launch task that uses another approach is superfluous.  Instead, there should be a launch task to attach to the FreeCAD running the VS Code debugger protocol.

Furthermore, having the task launch FreeCAD means that FreeCAD cannot also, simultaneously be executing via the C++ debugger.

**Comment by Pesc0** [comment]
I think there's a misunderstanding: when I click debug (F5) this is what happens:
- triggers a build
- launches gdb
- launches the python debugger, but it pauses in the prelaunchtask `waitfordebugpy`
- freecad launches and immediately executes the macro. this is the exact same thing as manually clicking Macro > attach to remote debugger > vscode > ok.
- `waitfordebugpy` notices the connection is available and succeeds. 
- the python debugger attaches to FC

Result: with one click i have both debuggers up and running. The scripts are just to automate this manual procedure. See video demonstration. 

https://github.com/FreeCAD/FreeCAD/assets/54720203/98fd5eca-8430-47d8-975f-7d1f0871e4c2

**Comment by oursland** [comment]
I understand now.  I withdraw my concern.

This looks good to me.

**Comment by Pesc0** [comment]
Ah looks like the video is broken, sorry. If everything is good I would merge, we can always tweak and improve things down the line.

**Pull request merged**